#### Import modul pandas

In [1]:
import pandas as pd

#### Import dataset

In [2]:
df = pd.read_csv("file_dataset.csv")

df.head()

Date      Time  Latitude  Longitude        Type  Depth  Depth Error  \
0  01/02/1965  13:44:18    19.246    145.616  Earthquake  131.6          NaN   
1  01/04/1965  11:29:49     1.863    127.352  Earthquake   80.0          NaN   
2  01/05/1965  18:05:58   -20.579   -173.972  Earthquake   20.0          NaN   
3  01/08/1965  18:49:43   -59.076    -23.557  Earthquake   15.0          NaN   
4  01/09/1965  13:32:50    11.938    126.427  Earthquake   15.0          NaN   

   Depth Seismic Stations  Magnitude Magnitude Type  ...  \
0                     NaN        6.0             MW  ...   
1                     NaN        5.8             MW  ...   
2                     NaN        6.2             MW  ...   
3                     NaN        5.8             MW  ...   
4                     NaN        5.8             MW  ...   

   Magnitude Seismic Stations  Azimuthal Gap  Horizontal Distance  \
0                         NaN            NaN                  NaN   
1                         NaN            NaN                  NaN   
2                         NaN            NaN                  NaN   
3                         NaN            NaN                  NaN   
4                         NaN            NaN                  NaN   

   Horizontal Error  Root Mean Square            ID  Source Location Source  \
0               NaN               NaN  ISCGEM860706  ISCGEM          ISCGEM   
1               NaN               NaN  ISCGEM860737  ISCGEM          ISCGEM   
2               NaN               NaN  ISCGEM860762  ISCGEM          ISCGEM   
3               NaN               NaN  ISCGEM860856  ISCGEM          ISCGEM   
4               NaN               NaN  ISCGEM860890  ISCGEM          ISCGEM   

  Magnitude Source     Status  
0           ISCGEM  Automatic  
1           ISCGEM  Automatic  
2           ISCGEM  Automatic  
3           ISCGEM  Automatic  
4           ISCGEM  Automatic  

[5 rows x 21 columns]

#### Informasi dataset

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23412 entries, 0 to 23411
Data columns (total 21 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Date                        23412 non-null  object 
 1   Time                        23412 non-null  object 
 2   Latitude                    23412 non-null  float64
 3   Longitude                   23412 non-null  float64
 4   Type                        23412 non-null  object 
 5   Depth                       23412 non-null  float64
 6   Depth Error                 4461 non-null   float64
 7   Depth Seismic Stations      7097 non-null   float64
 8   Magnitude                   23412 non-null  float64
 9   Magnitude Type              23409 non-null  object 
 10  Magnitude Error             327 non-null    float64
 11  Magnitude Seismic Stations  2564 non-null   float64
 12  Azimuthal Gap               7299 non-null   float64
 13  Horizontal Distance         160

### Data Cleaning

#### Cek NULL seluruh attribute

In [4]:
df.isna().sum()

Date                              0
Time                              0
Latitude                          0
Longitude                         0
Type                              0
Depth                             0
Depth Error                   18951
Depth Seismic Stations        16315
Magnitude                         0
Magnitude Type                    3
Magnitude Error               23085
Magnitude Seismic Stations    20848
Azimuthal Gap                 16113
Horizontal Distance           21808
Horizontal Error              22256
Root Mean Square               6060
ID                                0
Source                            0
Location Source                   0
Magnitude Source                  0
Status                            0
dtype: int64

#### Atasi NULL pada tiap attribute

In [5]:
from sklearn.impute import SimpleImputer

kolom_angka = df.select_dtypes(include=["int64", "float64"]).columns
kolom_string = df.select_dtypes(include=["object"]).columns

# atasi NULL pada attribute angka
df[kolom_angka] = SimpleImputer(strategy="mean").fit_transform(df[kolom_angka])

# # atasi NULL pada attribute string
df[kolom_string] = SimpleImputer(strategy="most_frequent").fit_transform(df[kolom_string])


#### Cek NULL setelah cleaning

In [6]:
df.isna().sum()

Date                          0
Time                          0
Latitude                      0
Longitude                     0
Type                          0
Depth                         0
Depth Error                   0
Depth Seismic Stations        0
Magnitude                     0
Magnitude Type                0
Magnitude Error               0
Magnitude Seismic Stations    0
Azimuthal Gap                 0
Horizontal Distance           0
Horizontal Error              0
Root Mean Square              0
ID                            0
Source                        0
Location Source               0
Magnitude Source              0
Status                        0
dtype: int64

#### Cek record duplikat

In [7]:
df[df.duplicated()]

Empty DataFrame
Columns: [Date, Time, Latitude, Longitude, Type, Depth, Depth Error, Depth Seismic Stations, Magnitude, Magnitude Type, Magnitude Error, Magnitude Seismic Stations, Azimuthal Gap, Horizontal Distance, Horizontal Error, Root Mean Square, ID, Source, Location Source, Magnitude Source, Status]
Index: []

[0 rows x 21 columns]

#### Atasi record duplikat

In [8]:
df.drop_duplicates(inplace=True)

#### Cek duplikat setelah cleaning

In [9]:
df[df.duplicated()]

Empty DataFrame
Columns: [Date, Time, Latitude, Longitude, Type, Depth, Depth Error, Depth Seismic Stations, Magnitude, Magnitude Type, Magnitude Error, Magnitude Seismic Stations, Azimuthal Gap, Horizontal Distance, Horizontal Error, Root Mean Square, ID, Source, Location Source, Magnitude Source, Status]
Index: []

[0 rows x 21 columns]

#### Simpan dataset yang telah di-cleaning

In [10]:
df.to_csv("file_CSV.csv")

# ____________________ SUPERVISED LEARNING ____________________

#### Import dataset yang akan digunakan untuk Supervised Learning

In [11]:
df = pd.read_csv("file_CSV.csv")

df.head()

Unnamed: 0        Date      Time  Latitude  Longitude        Type  Depth  \
0           0  01/02/1965  13:44:18    19.246    145.616  Earthquake  131.6   
1           1  01/04/1965  11:29:49     1.863    127.352  Earthquake   80.0   
2           2  01/05/1965  18:05:58   -20.579   -173.972  Earthquake   20.0   
3           3  01/08/1965  18:49:43   -59.076    -23.557  Earthquake   15.0   
4           4  01/09/1965  13:32:50    11.938    126.427  Earthquake   15.0   

   Depth Error  Depth Seismic Stations  Magnitude  ...  \
0     4.993115              275.364098        6.0  ...   
1     4.993115              275.364098        5.8  ...   
2     4.993115              275.364098        6.2  ...   
3     4.993115              275.364098        5.8  ...   
4     4.993115              275.364098        5.8  ...   

  Magnitude Seismic Stations  Azimuthal Gap  Horizontal Distance  \
0                  48.944618      44.163532              3.99266   
1                  48.944618      44.163532              3.99266   
2                  48.944618      44.163532              3.99266   
3                  48.944618      44.163532              3.99266   
4                  48.944618      44.163532              3.99266   

   Horizontal Error  Root Mean Square            ID  Source Location Source  \
0          7.662759          1.022784  ISCGEM860706  ISCGEM          ISCGEM   
1          7.662759          1.022784  ISCGEM860737  ISCGEM          ISCGEM   
2          7.662759          1.022784  ISCGEM860762  ISCGEM          ISCGEM   
3          7.662759          1.022784  ISCGEM860856  ISCGEM          ISCGEM   
4          7.662759          1.022784  ISCGEM860890  ISCGEM          ISCGEM   

  Magnitude Source     Status  
0           ISCGEM  Automatic  
1           ISCGEM  Automatic  
2           ISCGEM  Automatic  
3           ISCGEM  Automatic  
4           ISCGEM  Automatic  

[5 rows x 22 columns]

### Normalisasi Data

In [12]:
from sklearn.preprocessing import MinMaxScaler

df[kolom_angka] = MinMaxScaler().fit_transform(df[kolom_angka])

df[kolom_angka].head()

Latitude  Longitude     Depth  Depth Error  Depth Seismic Stations  \
0  0.590649   0.904493  0.189274     0.054692                0.294822   
1  0.484060   0.853759  0.115675     0.054692                0.294822   
2  0.346451   0.016736  0.030096     0.054692                0.294822   
3  0.110396   0.434562  0.022964     0.054692                0.294822   
4  0.545838   0.851190  0.022964     0.054692                0.294822   

   Magnitude  Magnitude Error  Magnitude Seismic Stations  Azimuthal Gap  \
0   0.138889          0.17517                    0.059616       0.122676   
1   0.083333          0.17517                    0.059616       0.122676   
2   0.194444          0.17517                    0.059616       0.122676   
3   0.083333          0.17517                    0.059616       0.122676   
4   0.083333          0.17517                    0.059616       0.122676   

   Horizontal Distance  Horizontal Error  Root Mean Square  
0             0.105313          0.076609          0.297321  
1             0.105313          0.076609          0.297321  
2             0.105313          0.076609          0.297321  
3             0.105313          0.076609          0.297321  
4             0.105313          0.076609          0.297321

### Data Spliting

In [13]:
from sklearn.model_selection import train_test_split


X = df[["Latitude", "Longitude", "Depth", "Depth Error",
        "Depth Seismic Stations", "Magnitude", "Magnitude Error",
        "Magnitude Seismic Stations", "Azimuthal Gap", "Horizontal Distance",
        "Horizontal Error", "Root Mean Square"]]
y = df["Location Source"]  # dependent

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=0)


# KNN (K-Nearest Neighbor)

In [14]:
from sklearn.neighbors import KNeighborsClassifier

model_KNN = KNeighborsClassifier(n_neighbors=5).fit(X_train, y_train)

model_KNN

KNeighborsClassifier()

#### Akurasi model KNN pada data testing

In [15]:
model_KNN.score(X_test, y_test)

0.8539709649871904

#### Akurasi model KNN pada data training

In [16]:
model_KNN.score(X_train, y_train)

0.9046511627906977

# Decision Tree

In [17]:
from sklearn.tree import DecisionTreeClassifier

model_DTree = DecisionTreeClassifier().fit(X_train, y_train)

model_DTree

DecisionTreeClassifier()

#### Akurasi model Decision Tree pada data testing

In [18]:
model_DTree.score(X_test, y_test)

0.9329632792485055

#### Akurasi model Decision Tree pada data training

In [19]:
model_DTree.score(X_train, y_train)

1.0

# Naive Bayes

In [20]:
from sklearn.naive_bayes import GaussianNB

model_GNB = GaussianNB().fit(X_train, y_train)

model_GNB

GaussianNB()

#### Akurasi model Naive Bayes pada data testing

In [21]:
model_GNB.score(X_test, y_test)

0.7860802732707088

#### Akurasi model Naive Bayes pada data training

In [22]:
model_GNB.score(X_train, y_train)

0.7769814902705268

# Support Vector Machine (SVM)

In [23]:
from sklearn.svm import SVC

model_SVM = SVC(kernel="linear").fit(X_train, y_train)

model_SVM

SVC(kernel='linear')

#### Akurasi model SVM pada data testing

In [24]:
model_SVM.score(X_test, y_test)

0.8689154568744663

#### Akurasi model SVM pada data training

In [25]:
model_SVM.score(X_train, y_train)

0.8736592311343142

# KESIMPULAN
Setelah dilakukan classification, didapatkan score: <br>
Model KNN = 85.39% <br>
Model Decision Tree = 93.29% <br>
Model Naive Bayes = 78% <br>
Model SVM = 86% <br>
Sehingga dapat disimpulkan, algoritma yang paling cocok diterapkan untuk dataset ini adalah <b>Decision Tree</b>